analyzers/src/SonarAnalyzer.CSharp/Rules/CloudNative/AzureFunctionsLogFailures.cs

```diff
@@ -0,0 +1,189 @@
+        {
+            return node.Ancestors().OfType<MethodDeclarationSyntax>().FirstOrDefault() is { } method
+                && method.AttributeLists.Any() // FunctionName has [AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)] so there must be an attribute
+                && model.GetDeclaredSymbol(method, cancellationToken) is { } methodSymbol
+                && model.Compilation.GetTypeByMetadataName(KnownType.Microsoft_Azure_WebJobs_FunctionNameAttribute.TypeName) is { } functionNameAttribute
+                && methodSymbol.GetAttributes().Any(a => a.AttributeClass.Equals(functionNameAttribute))
+                ? methodSymbol
+                : null;
+        }
+
+        private sealed class LoggerCallWalker : SafeCSharpSyntaxWalker
+        {
+            private static readonly int[] ValidLogLevel = new[]
+            {
+                3, // Warning
+                4, // Error
+                5, // Critical
+            };
+
+            private readonly ImmutableHashSet<ExpressionSyntax>.Builder invalidIvocationsBuilder = ImmutableHashSet.CreateBuilder<ExpressionSyntax>();
+
+            public LoggerCallWalker(ILanguageFacade languageFacade, SemanticModel model, ITypeSymbol iLoggerSymbol, CancellationToken cancellationToken)
+            {
+                Language = languageFacade;
+                Model = model;
+                ILogger = iLoggerSymbol;
+                CancellationToken = cancellationToken;
+                LoggerExtensions = new Lazy<INamedTypeSymbol>(() => Model.Compilation.GetTypeByMetadataName(KnownType.Microsoft_Extensions_Logging_LoggerExtensions.TypeName));
+                ILogger_Log = new Lazy<IMethodSymbol>(() => ILogger.GetMembers().OfType<IMethodSymbol>().FirstOrDefault(x => x.Name == "Log"));
+            }
+
+            private ILanguageFacade Language { get; }
+            private SemanticModel Model { get; }
+            private ITypeSymbol ILogger { get; }
+            private CancellationToken CancellationToken { get; }
+            private Lazy<INamedTypeSymbol> LoggerExtensions { get; }
+            private Lazy<IMethodSymbol> ILogger_Log { get; }
+            public bool HasValidLoggerCall { get; private set; }
+            public ImmutableArray<ExpressionSyntax> InvalidLoggerInvocations => invalidIvocationsBuilder.ToImmutableArray();
+
+            public override void Visit(SyntaxNode node)
+            {
+                CancellationToken.ThrowIfCancellationRequested();
+                if (HasValidLoggerCall)
+                {
+                    return;
+                }
+                base.Visit(node);
+            }
+
+            public override void VisitInvocationExpression(InvocationExpressionSyntax node)
+            {
+                if (Model.GetSymbolInfo(node, CancellationToken).Symbol is IMethodSymbol { ReceiverType: { } receiver }
+                    && receiver.DerivesOrImplements(ILogger))
+                {
```

Comment:
I re-factored that method a bit and we should discuss this with the latest version. I implemented it in a way to avoid FP here. I think it is fairly common to have your own wrapped extension methods to e.g. amend your logging message. The scaffolded VS solution brings its own set of logger extensions already:

[LogFunctionRetryAttempt](https://github.com/Azure/azure-webjobs-sdk/blob/2d73523800392fad588b068dbac4bab5214f68d3/src/Microsoft.Azure.WebJobs.Host/Loggers/Logger/LoggerExtensions.cs#L243-L246)
[LogMetric](https://github.com/Azure/azure-webjobs-sdk/blob/2d73523800392fad588b068dbac4bab5214f68d3/src/Microsoft.Azure.WebJobs.Host/Loggers/Logger/LoggerExtensions.cs#L184-L200)

Reply:
Do we have a UT for the `Custom` class there? I might have missed that.